### DISCUSSION TITLE
Some problems of neural networks

### DISCUSSION THREAD

**Discussion by turbulence-vortex** [mdolab/dafoam#969] [General]
Dear professor He：
I have some problems:
1.When we come across: `Warning! output values are bounded between...` When doing this, may I ask if the "bound" operation is included in the gradient of the backpropagation of dafoam? Since we want the value of beta to be as close as possible to the range we have set, rather than having a wide output range, I wonder if you have any good suggestions to avoid this issue occurring during the process.
2.DAFom seems to use the method of inputting neural networks by individual grid cells, right? It is equivalent to batch size = 1. I think this might be a bit slow in large neural networks. Could we pass it in using a matrix format instead?
3.I have employed the coupled FI and ML as well as the decoupled FI and ML approaches for the same problem. The latter can better align with my goals, while the objective loss function of the former fails to decrease effectively. I'm not quite sure why this kind of experience occurred. I have tried changing the initial seed number and increasing the number of layers in the neural network, but neither approach has worked well. I wonder if your team has any good solutions when conducting the tests?

**Comment by turbulence-vortex** [comment]
I wonder if your team has conducted any tests related to neural networks on the airfoil. I have noticed that it is extremely difficult to obtain satisfactory results on the airfoil. For instance, the field tends to become uniform, and the distribution of beta does not resemble the result obtained through decoupling.

**Comment by friedenhe** [comment]
1. If "Warning! output values are bounded between..." happens, I think we should consider the flow solution failing, and will not use the results in optimization. So the gradient will NOT backpropagate through this case.
2. Yes, the neural network is used to predict beta for EACH cell. But the batch size is not 1. Check: https://github.com/mdolab/dafoam/blob/31433b41d1d59638b459d24f02c7f89a756848c2/dafoam/pyDAFoam.py#L2370. We can prescribe a large batch size for fast prediction.
3. Yes, the decoupled FI and ML should make the FI converge much better than the coupled one. If the case does not converge well, first check if the flow and adjoint converge. You can post your log file, and I can take a look. It is hard to tell what went wrong because it is very case-dependent.
4. Yes, we did FI for airfoils using DAFoam. Check this [paper](https://arxiv.org/abs/2511.18276)
